b2sdk/raw_simulator.py

```diff
@@ -1079,33 +1085,46 @@ def upload_file(
         self, upload_url, upload_auth_token, file_name, content_length, content_type, content_sha1,
         file_infos, data_stream
     ):
-        assert upload_url == upload_auth_token
-        url_match = re.match(r'https://upload.example.com/([^/]*)/([^/]*)', upload_url)
-        if url_match is None:
-            raise BadUploadUrl(upload_url)
-        if len(self.upload_errors) != 0:
-            raise self.upload_errors.pop(0)
-        bucket_id, upload_id = url_match.groups()
-        bucket = self._get_bucket_by_id(bucket_id)
-        response = bucket.upload_file(
-            upload_id, upload_auth_token, file_name, content_length, content_type, content_sha1,
-            file_infos, data_stream
-        )  # yapf: disable
-        file_id = response['fileId']
-        self.file_id_to_bucket_id[file_id] = bucket_id
+        if not self.currently_used_auth_tokens[upload_auth_token].acquire(False):
+            raise UploadTokenUsedConcurrently(upload_auth_token)
+
+        try:
+            assert upload_url == upload_auth_token
+            url_match = re.match(r'https://upload.example.com/([^/]*)/([^/]*)', upload_url)
+            if url_match is None:
+                raise BadUploadUrl(upload_url)
+            if len(self.upload_errors) != 0:
+                raise self.upload_errors.pop(0)
+            bucket_id, upload_id = url_match.groups()
+            bucket = self._get_bucket_by_id(bucket_id)
+            response = bucket.upload_file(
+                upload_id, upload_auth_token, file_name, content_length, content_type, content_sha1,
+                file_infos, data_stream
+            )  # yapf: disable
+            file_id = response['fileId']
+            self.file_id_to_bucket_id[file_id] = bucket_id
+        finally:
+            self.currently_used_auth_tokens[upload_auth_token].release()
         return response
 
     @set_token_type(TokenType.UPLOAD_PART)
     def upload_part(
         self, upload_url, upload_auth_token, part_number, content_length, sha1_sum, input_stream
     ):
-        url_match = self.UPLOAD_PART_MATCHER.match(upload_url)
-        if url_match is None:
-            raise BadUploadUrl(upload_url)
-        file_id = url_match.group(1)
-        bucket_id = self.file_id_to_bucket_id[file_id]
-        bucket = self._get_bucket_by_id(bucket_id)
-        return bucket.upload_part(file_id, part_number, content_length, sha1_sum, input_stream)
+        if not self.currently_used_auth_tokens[upload_auth_token].acquire(False):
```

Comment:
same here with the context manager

Reply:
@ppolewicz fixed in https://github.com/reef-technologies/b2-sdk-python/pull/53/commits/1839cc08fabff101b01b74624fa374953078c9a2